---
title: "integration of CT an RA cells"
author: "Babacar NDAO"
date: "21/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

A- load 10X data generated by cellranger count

## -----------------------------------------------------------------------------
## Libraries
## -----------------------------------------------------------------------------
```{r}
suppressMessages(library("getopt"))
suppressMessages(library("Matrix"))
suppressMessages(library("Seurat"))

#source("R_src/funForLoading.R")
```

```{r}
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/funForLoading.R")
```



## -----------------------------------------------------------------------------
## Command line args
## -----------------------------------------------------------------------------

```{r}
spec = matrix(c(
  'help',        'h', 0, "logical",   "Help about the program",
  'inputMatrixDir', 'b', 1, "character", "REQUIRED:10X expression data (matrix files directory path)",
  'outfile',     'f',1, "character", 'Output file path (default .10X_data.rds)',
  'cellranger',   'c',1, "character", "cellranger version used to generate matrix files (default 2)",
  "subSample",    "p",1, "character", "proportion of cell to subsample (by default all the cells are used (e.g. 0.5))",
  "sampleInfo",   "s", 1, "character", "sample information that will be added to pData with the following format: age=2_months,runDate=10_12_2017..",
  "sampleName",   "n", 1, "character", "if matrix file have the sampl name as prefix (name_matrix.mtx...) provide it here"
  
), byrow=TRUE, ncol=5);

opt = getopt(spec)
```

# if help was asked, print a friendly message
# and exit with a non-zero error code
# For test
# opt <- list()
```{r}
opt$inputMatrixDir <- "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/input/RAW/PLZF_RARA_RA/"
opt$cellranger <- 4
opt$sampleInfo <- "age=Adult,runDate=11/2020,sampleName=RNA_RA"
opt$outfile <- "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/input/10X_data_RNA_RA.rds"

args <- commandArgs()
if ( !is.null(opt$help) | is.null(opt$inputMatrixDir)) {
  cat("loading 10X genomics single cells data and save it as a seurat R object.")
  cat(getopt(spec, usage=TRUE))
  q(status=1)
}
```

## -----------------------------------------------------------------------------
## Processing data
## -----------------------------------------------------------------------------

# output

```{r}
if(is.null(opt$outfile)) {
  opt$outfile <- "10X_data.rds"
  outdir = './'
} else {
  outdir <- paste(strsplit(opt$outfile,"/")[[1]][-length(strsplit(opt$outfile,"/")[[1]])],collapse="/")
  dir.create(outdir, showWarnings=FALSE)
}


if(is.null(opt$cellranger)) {
  opt$cellranger <- "2"
}


if (opt$cellranger == "3" | opt$cellranger == "4") {
  gbm <- loadCellRangerMatrix_cellranger3(opt$inputMatrixDir,sample_name=opt$sampleName)
} else
{
  gbm <- loadCellRangerMatrix(opt$inputMatrixDir,sample_name=opt$sampleName)
}


fd <- gbm$fd
pd <- gbm$pd

# Column 'symbol' is the one (from cellRanger workflow) that corresponds to featureData's gene short names.

colnames(fd)[which(colnames(fd)=="symbol")]<- "gene_short_name"

for (c in c(1:length(colnames(fd)))){
  if (is.na(colnames(fd)[c])) {
    colnames(fd)[c] <- paste0("col_",c)
  }
}


seurat <- CreateSeuratObject(counts = gbm$exprs,
                             assay = "RNA",
                             meta.data = gbm$pd)

# ## Subsample ## DEPRECATED
# if(is.null(opt$subSample)==F) {
#   opt$subSample <- as.numeric(opt$subSample)
#   print(opt$subSample)
#   
#   cellSubset <- sample(rownames(pData(gbm_cds)),size = opt$subSample*length(rownames(pData(gbm_cds))))
#   gbm_cds <- gbm_cds[,cellSubset] 
# }
```

#Add sample infos

```{r}
print(opt$sampleInfo)
sampleInfos <- strsplit(opt$sampleInfo,split=",")[[1]]
for (i in sampleInfos) {
  print(i)
  info <- strsplit(i,split = "=")[[1]][1]
  print(info)
  value <- strsplit(i,split = "=")[[1]][2]
  print(value)
  seurat@meta.data[,info] <- value
}
```

# save seurat
```{r}
output <- list(seurat = seurat,featureData = fd)
#saveRDS(output,file = opt$outfile)
```



## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------

B- prepare data for Seurat Object

#Load required packages
```{r}
suppressMessages(library(Seurat))
suppressMessages(library(scran))
suppressMessages(library(getopt))
suppressMessages(library(grid))
suppressMessages(library(gridExtra))
suppressMessages(library(scales))
suppressMessages(library(ggplot2))
```

#load homemade function
```{r}
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/data_preparation.R")
```

# Quality control and preparation of 10X data for analysis

## -----------------------------------------------------------------------------
## Command line args
## -----------------------------------------------------------------------------

```{r}
spec = matrix(c(
  'help',        'h', 0, "logical",   "Help about the program",
  'inputRDS',  'i', 1, "character", "REQUIRED: 10X data prepared as a list of seurat object and feature data table (.RDS generated by prepare_data.R).",
  "mitochFilter", "m",1, "numeric", "mitochondrial protein gene RNA percentage filtering (eg 10 to remove cells with more than 10% of mitoch protein transcripts, default none)",
  'outdir',     'o',1, "character", 'Outdir path (default ./)',
  'upperMad',    'u', 1, "numeric", "cells with more than u mads log counts will be discard (default 2)",
  "lowerMad",   'l', 1, "numeric", "cells with less than l mads log counts will be discard (default 2)"
), byrow=TRUE, ncol=5);

opt = getopt(spec)
```

## directories and options

```{r}
inputRDS <- "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/input/10X_data_RNA_Ctrl.rds"
outdir <- "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/output/RNA/PLZF_RARA_CT/dataPreparation"


if (is.null(opt$outdir)) {
  opt$outdir = "./"
}

if (is.null(opt$mitochFilter)) {
  opt$mitochFilter <- 5
}

if (is.null(opt$upperMad)) {
  opt$upperMad <- 2
}

if (is.null(opt$lowerMad)) {
  opt$lowerMad <- 3
}


input <- readRDS(inputRDS)
seurat <- input$seurat
fd <- input$featureData
```

# First getting cell cycle phases with cyclone function from package scran (see data_preparation.R)
```{r}
seurat <- getCellCyclePhasesSeurat(seurat,outdir = paste(outdir,"/cellCycle",sep =""))
```


# Change ensembl gene names in the seurat object ot gene short names to be able to use PercentageFeatureSet Seurat function for mt and rb transcript proportion
#check for dup genes

# Only needed if ensemble id (it is the case in the workflow)
```{r}
dupGeneNames <- fd[which((duplicated(fd$gene_short_name))),"gene_short_name"]

if(length(dupGeneNames) == 0) {
  rownames(seurat) <- fData(gbm_to_seurat)$gene_short_name
  
} else {
  #write.csv(fd[which(is.element(fd$gene_short_name,dupGeneNames)),],
  #          paste(opt$outdir,"/dupGenesName.csv",sep = "")
  #)
  print("Dup gene short names existing, making them unique...")
  newSeuratCount <- as.matrix(GetAssayData(seurat,slot = 'counts',assay = "RNA"))
  rownames(newSeuratCount) <- make.unique(fd$gene_short_name, sep = "--")
  seurat <- CreateSeuratObject(counts = newSeuratCount,
                               assay = "RNA",
                               meta.data = seurat@meta.data
                               )

}
```

# Compute percent Rb and percent mt
```{r}
seurat$percentMito <- PercentageFeatureSet(
  seurat,
  pattern = 'mt-',
  assay = 'RNA'
)


seurat$percentRibo <- PercentageFeatureSet(
  seurat,
  pattern = 'Rps|Rpl|Mrp', #it might be better to have the exact gene set through GO...
  assay = 'RNA'
)
```

# Filtering cells outside two MADs from median
```{r}
print("Matrix dim before cell filtering :")
dim(seurat)

# plot qc metrics before filtering
png(paste(outdir,"/qcMetricsRaw.png",sep = ""))
VlnPlot(seurat, features = c("nFeature_RNA", "nCount_RNA", "percentMito","percentRibo"), ncol = 3,pt.size = 0.001)
dev.off()

# plot some correlation between qc metrics 
plot1 <- FeatureScatter(seurat, feature1 = "nCount_RNA", feature2 = "percentMito")
plot2 <- FeatureScatter(seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

png(paste(outdir,"/qcMetricsCr.png",sep = ""))
grid.arrange(plot1,plot2)
dev.off()

minCountThreshold <- exp(median(log(seurat@meta.data$nCount_RNA)) - opt$lowerMad*mad(log(seurat@meta.data$nCount_RNA)))
maxCountThreshold <- exp(median(log(seurat@meta.data$nCount_RNA)) + opt$upperMad*mad(log(seurat@meta.data$nCount_RNA)))

png(paste(outdir,"/cutOffLogCount.png",sep = ""))
VlnPlot(seurat, features = c( "nCount_RNA"),pt.size = 0.001) + 
  geom_hline(yintercept = minCountThreshold) +
  geom_hline(yintercept = maxCountThreshold)
dev.off()

seurat <- subset(seurat, subset = nCount_RNA > minCountThreshold &
                   nCount_RNA < maxCountThreshold & 
                   percentMito < opt$mitochFilter)

#Check if some cells don't have an assinged cell cycles phases 
table(!is.na(seurat@meta.data$phases))
png(paste(outdir,"/assignedPhases.png",sep = ""))
barplot(table(!is.na(seurat@meta.data$phases)),main="Phases assigned")
dev.off()

if("FALSE" %in% names(table(!is.na(seurat@meta.data$phases)))) {
  print("Discard cells with no cell cycle phases assinged (too few genes")
  print(table(!is.na(seurat@meta.data$phases))[1])
  seurat@meta.data$phases[is.na(seurat@meta.data$phases)] <- "noAssigned"
  Idents(seurat) <- "phases"
  seurat <- subset(seurat, idents = "noAssigned",invert = T)
  Idents(seurat) <- "orig.ident"
  
}

png(paste(outdir,"/ccPhases.png",sep = ""))
barplot(table(seurat@meta.data$phases),main="Phases")
dev.off()

# plot qc metrics after filtering
png(paste(outdir,"/qcMetricsFiltered.png",sep = ""))
VlnPlot(seurat, features = c("nFeature_RNA", "nCount_RNA", "percentMito","percentRibo"), ncol = 3,pt.size = 0.001)
dev.off()

print("Matrix dim after cell filtering :")
print(dim(seurat))

```

# save seurat
```{r}
#saveRDS(object = seurat,file = paste(outdir,"/seurat_treated.rds",sep =""))
```



## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------

C- Seurat4


## Libraries
## -----------------------------------------------------------------------------

```{r}
suppressMessages(library(Seurat))
suppressMessages(library(BiocParallel))
suppressMessages(library(getopt))
suppressMessages(library(gridExtra))
#suppressMessages(library(gProfileR))
suppressMessages(library(RColorBrewer))
suppressMessages(library(readxl))
suppressMessages(library(stringr))
suppressMessages(library(scales))
suppressMessages(library(grid))
suppressMessages(library(ggplot2))

```

#load homemade functions
```{r}
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/getSigPerCells.R")
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/Enrichment.R")
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/funForSeurat.R")
```

# Seurat 3 analysis of an individual sample required for using CaSTLe script.

## -----------------------------------------------------------------------------
## Command line args
## -----------------------------------------------------------------------------

```{r}
spec = matrix(c(
  'help',        'h', 0, "logical",   "Help about the program",
  'inputRDS',  'i', 1, "character", "REQUIRED : 10X data prepared as monocle or seurat object (.RDS generated by prepare_data.R).",
  'outdir',     'o',1, "character", 'Outdir path (default ./)',
  "num_dim",      'n',1,"numeric", "Number of dimension to use for ordering (eg n first pc of PCA on input data) 10 by default",
  "correction",      "b",1,"character", "Covariable to use as blocking factor (eg one or several columns of pData: betch, cell cycle phases... separated by +)",
  "minPropCellExp", "p",1,"numeric", "minimal proportion of cells that expressed the genes kept for the analaysis 0.001 by default",
  "norm_method",    "z",1, "character", "normalisation method, logNorm seurat (by default) or sctransform",
  "resolution", "r", 1,"numeric", "resolution for Seurat clustering 0.9 by default",
  "signaturesFile", "s", 1, "character", "path to folder with signature list store as rds object",
  "identRemoved",    "d", 1, "character", "Optionnal cluster to remove only work if input is a seurat object",
  "nonExpressedGenesRemoved", "e", 0,"logical", "non expressed gene already removed default to FALSE",
  "gprofiler", "g", 0, "logical", "if true doing gprofiler default to true",
  "logfc_threshold", "l", 1, "numeric", "logfc threshold for finding cluster markers (1 cluster vs all deg) 0.25 by default"
  
), byrow=TRUE, ncol=5);

opt = getopt(spec)
```

## input, output directories and options
```{r}
## For testing
# 
# opt <- list()
outdir <- "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/output/RNA/PLZF_RARA_RA/Seurat4"
correction <- "G2M_score+S_score+G1_score"
inputRDS <- "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/output/RNA/PLZF_RARA_RA/dataPreparation/seurat_treated.rds"


#set default arguments values

if (is.null(opt$num_dim)) {
  opt$num_dim <- 15
}

if (is.null(opt$resolution)) {
  opt$resolution <- 0.6
}

if (is.null(opt$minPropCellExp)) {
  opt$minPropCellExp <- 0.001
  print(opt$minPropCellExp)
}


if (is.null(opt$logfc_threshold)) {
  opt$logfc_threshold = 0.25
} 

print(paste("logfc threshold:", opt$logfc_threshold) )

if (is.null(opt$outdir)) {
  opt$outdir = "./"
}

if (is.null(opt$norm_method)) {
  opt$norm_method = "logNorm"
}



if (is.null(opt$nonExpressedGenesRemoved)) {
  opt$nonExpressedGenesRemoved = F
} else {
  opt$nonExpressedGenesRemoved = T
}

# if (is.null(opt$gprofiler)) {
#   opt$gprofiler <- T
# } else {
#   opt$gprofiler <- F
# }


# get correction vector
if (!is.null(correction)) {
  corrections <- strsplit(x = correction,split = "\\+")[[1]]
} else {
  corrections <- NULL
}



dir.create(outdir,recursive = T,showWarnings = F)

print(opt)

# Loading
seurat <- readRDS(inputRDS)
```

# Remove non expressed genes if not done
```{r}
if(opt$nonExpressedGenesRemoved == F) {
  genesFiltered <- list(genes_before = dim(seurat)[1])
  
  print("remove non expressed genes (non expressed in at least X% of the cells X user option in monocle dp feature 5% in seurat tutorial 0,1%)")
  threshold <- opt$minPropCellExp * ncol(seurat)

  seurat <- CreateSeuratObject(counts = as.matrix(GetAssayData(seurat,slot = 'counts',assay = "RNA")),
                               assay = "RNA",
                               meta.data = seurat@meta.data,
                               min.cells = threshold)
  genesFiltered$genes_after <- dim(seurat)[1]
  
  png(paste(outdir,"/genesFiltering.png",sep = ""))
  barplot(unlist(genesFiltered),main = "Genes filtering")
  dev.off()
  
}
  
```

#################################################################################################
####################################### Seurat Workflow #########################################
#################################################################################################

```{r}
classicSeuratWorkflow <- function(seurat, corrections,outdir) {
  dir.create(outdir,recursive = T,showWarnings = F)
  
  if (opt$norm_method == "sctransform") {
    
    seurat <- SCTransform(object = seurat, vars.to.regress = corrections)
    
  } else {
    
    seurat <- NormalizeData(object = seurat)
    seurat <- FindVariableFeatures(object = seurat,selection.method = "vst", nfeatures = 2000, verbose = T)
    seurat <- ScaleData(object = seurat,vars.to.regress = corrections)
    
  }
  
  seurat <- RunPCA(object = seurat)
  
  
  png(paste(outdir,"/ElbowPlot.png",sep =""))
  ElbowPlot(object = seurat,ndims = 30)
  dev.off()
  
  
  print("Clustering...")
  
  seurat <- FindNeighbors(object = seurat,dims = c(1:opt$num_dim),k.param = 20)
  seurat <- FindClusters(object = seurat,resolution = c(0.5,0.6,0.7,0.8,0.9,1,1.2))
  
  print("Running TSNE...")
  
  seurat <- RunTSNE(seurat,dims = c(1:opt$num_dim))
  
  print("Running UMAP...")
  
  seurat <- RunUMAP(seurat,dims = c(1:opt$num_dim))
  
  print("UMAP ok")
  print(colnames(seurat@meta.data))
  
  colPrefix <- "RNA_snn_res."
  if(opt$norm_method == "sctransform") {
    colPrefix <- "SCT_snn_res."
  }
  
  umapListRes <- list()
  tsneListRes <- list()
  for (r in c(0.6,0.8,1,1.2)) {
    umapListRes[[as.character(r)]] <- DimPlot(seurat,
                                              reduction = "umap",
                                              label = T,
                                              group.by = paste(colPrefix,r,sep="")) + 
      NoLegend() +
      ggplot2::ggtitle(paste("res",r))
    
    tsneListRes[[as.character(r)]] <- DimPlot(seurat,
                                              reduction = "tsne",
                                              label = T,
                                              group.by = paste(colPrefix,r,sep="")) + 
      NoLegend() +
      ggplot2::ggtitle(paste("res",r))
  }
  
  png(paste(outdir,"/tsne_different_res.png",sep = ""),height = 800,width = 800)
  grid.arrange(tsneListRes[[1]],tsneListRes[[2]],tsneListRes[[3]],tsneListRes[[4]])
  dev.off()
  
  png(paste(outdir,"/umap_different_res.png",sep = ""),height = 800,width = 800)
  grid.arrange(umapListRes[[1]],umapListRes[[2]],umapListRes[[3]],umapListRes[[4]])
  dev.off()
  
  
  
  
  Idents(seurat) <- paste(colPrefix,opt$resolution,sep = "")
  seurat@meta.data$numclust <- seurat@meta.data[,paste(colPrefix,opt$resolution,sep = "")]
  
  
  #Check for unwanted source of variation
  pPhases <- DimPlot(seurat,group.by = "phases")
  if (!is.null(seurat@meta.data$predicted)) {
    pPred <- DimPlot(seurat,group.by = "predicted")
  } else {
    pPred <- DimPlot(seurat)
  }
  
  pUMI <- FeaturePlot(seurat, "nCount_RNA")
  pMito <- FeaturePlot(seurat, "percentMito")
  
  png(paste(outdir,"/umap_factors.png",sep = ""),height = 800,width = 800)
  grid.arrange(pPhases,pPred,pUMI,pMito)
  dev.off()
  
  #Check for unwanted source of variation
  pPhases <- DimPlot(seurat,group.by = "phases",reduction = "tsne")
  if (!is.null(seurat@meta.data$predicted)) {
    pPred <- DimPlot(seurat,group.by = "predicted",reduction = "tsne")
  } else {
    pPred <- DimPlot(seurat,reduction = "tsne")
  }
  pUMI <- FeaturePlot(seurat, "nCount_RNA",reduction = "tsne")
  pMito <- FeaturePlot(seurat, "percentMito",reduction = "tsne")
  
  png(paste(outdir,"/tsne_factors.png",sep = ""),height = 800,width = 800)
  grid.arrange(pPhases,pPred,pUMI,pMito)
  dev.off()
  
  return(seurat)
  
}
  
```

## correction before integration
```{r}
## First without integration without correction
unintegrated <- seurat
DefaultAssay(unintegrated) <- "RNA"
classicSeuratWorkflow(unintegrated,correction = NULL,outdir = paste0(outdir,"/SeuratWoIntegrationWoCr/"))

## without integration with correction

classicSeuratWorkflow(unintegrated,correction = corrections,outdir = paste0(outdir,"/SeuratWoIntegrationWithCr/"))


## With integration without correction
if (!is.null(corrections)) {
  print("Seurat workflow without correction")
  classicSeuratWorkflow(seurat,correction = NULL,outdir = paste0(outdir,"/SeuratWithoutCr/"))
}

## With integration with correction

seurat <- classicSeuratWorkflow(seurat,corrections = corrections,outdir =outdir)
```

#################################################################################################
####################################### Markers analysis ########################################
#################################################################################################

```{r}
markers <- FindAllMarkers(seurat,only.pos = T,logfc.threshold= opt$logfc_threshold)
markers <- markers[which(markers$p_val_adj < 0.05),]

write.table(x = markers,paste(outdir,"/markers.tsv", sep =""),sep = "\t",quote = F,row.names = F,col.names = T)


dir.create(paste(outdir,"/markers/",sep = ""))

ylab <- "LogNormalized UMI counts"
if (opt$norm_method == "sctransform") {
  ylab <- "Expression level"
}

for (numClust in unique(markers$cluster)) {
  print(head(markers[which(markers$cluster == numClust),],n=9))
  png(paste(outdir,"/markers/Cluster_",numClust,"_topGenesVlnPlot.png",sep =""),width = 1000, height = 1000)
  plot(VlnPlot(object = seurat, features = head(markers[which(markers$cluster == numClust),"gene"],n=9),pt.size = 0.5) + 
         labs(x = "Clusters",y= ylab,colour = "black") +
         theme(axis.text = element_text(size=20),
               plot.title = element_text(size=25)) )
  dev.off()
  
}



## Add signatures scores
dir.create(paste(outdir,"/cellSignatures/",sep = ""))
signatures <- readRDS(opt$signaturesFile)
names(signatures) <- paste0(names(signatures),"_",seurat$sampleName[1])

for (sig in c(1:length(signatures))) {
  sigName <- names(signatures)[sig]
  signature <- signatures[[sig]]
  seurat <- scoreCells3(seurat,signature,outdir= paste(outdir,"/cellSignatures/",sep=""),sigName)
}
```



#################################################################################################
##########                     Cluster summary table                                    #########
#################################################################################################


```{r}
clust_table <- data.frame()

print("Creating cluster summary table ")

  firstup <- function(x) {
    substr(x, 1, 1) <- toupper(substr(x, 1, 1))
    x
  }

  colnames(markers) <- firstup(colnames(markers))

getClustInfo <- function(clust,signatures,seurat,markers) {
  
  clustInfo <- list()
  clustInfo$num_cells <- dim(seurat@meta.data[which(seurat@active.ident==clust),])[1]
  clustInfo$percent_cells <- clustInfo$num_cells/dim(seurat@meta.data)[1]
  percentPhases <- table(seurat@meta.data[which(seurat@active.ident==clust),"phases"])/length(seurat@meta.data[which(seurat@active.ident==clust),"phases"]) #In fact this is fraction not percentage
  
  if(!is.null(seurat@meta.data$predicted)) {
    percentPredicted <- table(seurat@meta.data[which(seurat@active.ident==clust),"predicted"])/length(seurat@meta.data[which(seurat@active.ident==clust),"predicted"]) #In fact this is fraction not percentage
    
    for (p in unique(seurat@meta.data$predicted)) {
      print(p)
      if (is.element(p,names(percentPhases))) {
        clustInfo[[p]] <- percentPhases[p] 
      } else {
        clustInfo[[p]] <- 0
      }
    }
    
  }
  
  for (p in c("G1_G0","S","G2_M")) {
    if (is.element(p,names(percentPhases))) {
      clustInfo[[p]] <- percentPhases[p] 
    } else {
      clustInfo[[p]] <- 0
    }
  }
  
  
  clustInfo$median_genes_expressed <- median(seurat@meta.data[which(seurat@active.ident==clust),"nFeature_RNA"])
  clustInfo$median_nUMI <- median(seurat@meta.data[which(seurat@active.ident==clust),"nCount_RNA"])
  clustInfo$median_percentMitochGenes <- median(seurat@meta.data[which(seurat@active.ident==clust),"percentMito"])
  
  
  clustSig <- lapply(signatures,testHyperSig3,seurat,markers,clust) 
  
  clustInfo <- c(clustInfo,clustSig)
  
}

#allSignatures <- c(signatures,signaturesRodriguez)

clust_list <- lapply(levels(unique(seurat@active.ident)),getClustInfo,signatures,seurat,markers)

names(clust_list) <- paste("cluster_",levels(unique(seurat@active.ident)),sep="")

saveRDS(clust_list,paste(opt$outdir,"/clust_list_save.rds",sep =""))

clust_table <- as.data.frame(matrix(unlist(clust_list), nrow=length(unlist(clust_list[1]))))
colnames(clust_table) <- names(clust_list)
rownames(clust_table) <- names(clust_list[[1]])


clust_df <- as.data.frame(t(clust_table))

write.table(x = clust_df,file = paste(outdir,"/clusters_table.tsv",sep =""),sep="\t",quote=F,col.names = NA)
```

# save seurat
```{r}
#saveRDS(seurat,file = paste(outdir,"/seurat.rds",sep = ""))
```



## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------
## -----------------------------------------------------------------------------

D- seurat4 integration of CT and RA


## Path to Rscript, for Rscript runing on cluster
#!/shared/ifbstor1/software/miniconda/envs/r-4.1.1/bin/Rscript

## Libraries
## -----------------------------------------------------------------------------

```{r}
#suppressMessages(library(monocle))
suppressMessages(library(Seurat))
suppressMessages(library(BiocParallel))
suppressMessages(library(getopt))
suppressMessages(library(gridExtra))
#suppressMessages(library(gProfileR))
suppressMessages(library(RColorBrewer))
suppressMessages(library(readxl))
suppressMessages(library(stringr))
suppressMessages(library(scales))
suppressMessages(library(cowplot))
suppressMessages(library(sctransform))
suppressMessages(library(stringr))
suppressMessages(library(plyr))
suppressMessages(library(grid))
suppressMessages(library(ggplot2))
```

#load homemade functions
```{r}
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/getSigPerCells.R")
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/Enrichment.R")
source("/shared/ifbstor1/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/R_src/funForSeurat.R")
```


# Seurat 3 integration workflow

## -----------------------------------------------------------------------------
## Command line args
## -----------------------------------------------------------------------------

```{r}
spec = matrix(c(
  'help',        'h', 0, "logical",   "Help about the program",
  'inputFiles',  'i', 1, "character", "REQUIRED: 10X dataset paths prepared as seurat object (.RDS generated by prepare_data.R) separated by +",
  'signaturesFile', 's',1, "character", "REQUIRED: signatures rds file",
  'outdir',     'o',1, "character", 'Outdir path (default ./)',
  "num_dim_CCA", 'q',1,"numeric", "First n dimensions of CCA to use for FindIntegrationAnchors Seurat function (15 by default)",
  "num_dim_weight", 'w',1,"numeric", "First n dimensions to use for IntegrateData Seurat function (15 by default)",
  "num_dim",      'n',1,"numeric", "First n dimensions of PCA to use for clustering, UMAP and TSNE (15 by default)",
  "num_dim_integrated",'N',1,"numeric", "Number of PCA dimension computed to analyse integrated data (40 by default)",
  "cores",        'c',1, "numeric", "Number of cores to use for ordering (for differencially expressed gene between clusters test)",
  "resolution",    'r',1, "numeric", "resolution for smp.combined clustering",
  "correction",      "b",1,"character", "Covariable to use as blocking factor (eg one or several columns of pData: betch, cell cycle phases... separated by +)",
  "logfc_threshold", "l", 1, "numeric", "logfc threshold for finding cluster markers (1 cluster vs all deg) 0.25 by default",
  "rodriguezSig", "k", 1, "character", "path for xls file of Rodriguez result",
  "norm_method", "z",1, "character", "normalization method, logNorm (by default) or sctransform",
  "reusePca", "p", 0, "character","re use pca calculated before when caculating anchor weights for each dataset default to FALSE (permit to correct for cell cycle before integration)"
), byrow=TRUE, ncol=5);

opt = getopt(spec)
```

## input, output directories and options
```{r}
inputFiles = "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/output/RNA/PLZF_RARA_CT/Seurat4/seurat.rds+/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/output/RNA/PLZF_RARA_RA/Seurat4/seurat.rds"
outdir = "/shared/projects/scrna_gmp_rara/stage_babacar/scranseq/analysis/output/RNA/Seurat4_integration"
correction <- "G2M_score+S_score+G1_score"


#set default arguments values

if (is.null(opt$resolution)) {
  opt$resolution <- 0.5
}

if (is.null(opt$logfc_threshold)) {
  opt$logfc_threshold <- 0.25
}

if (is.null(opt$num_dim_CCA)) {
  opt$num_dim_CCA <- 15
}

if (is.null(opt$num_dim_weight)) {
  opt$num_dim_weight <- 15
}

if (is.null(opt$num_dim)) {
  opt$num_dim <- 15
}

if (is.null(opt$num_dim_integrated)) {
  opt$num_dim_integrated <- 50
}

if (is.null(opt$norm_method)) {
  opt$norm_method <- "logNorm"
}

if (is.null(opt$reusePca)) {
  opt$reusePca <- FALSE
}

print(opt$reusePca)

dir.create(outdir,recursive = T,showWarnings = F)


corrections <- strsplit(x = correction,split = "\\+")[[1]]


dir.create(outdir,recursive = T,showWarnings = F)


smp.listFile <- strsplit(inputFiles,split = "\\+")[[1]]

smp.list <- list()
```

#load dataset,  
```{r}
for (i in 1:length(x = smp.listFile)) {
  #For testing, in final workflow sampleName will be incorporated in metadata with the loading of cell ranger matrix
  smp.list[[i]] <- readRDS(smp.listFile[i])
  sampleName <- unique(smp.list[[i]]@meta.data$sampleName)
  smp.list[[i]] <- RenameCells(smp.list[[i]],add.cell.id = sampleName)
  
  if(opt$norm_method != "sctransform") {
    smp.list[[i]] <- NormalizeData(object = smp.list[[i]], verbose = FALSE)
    smp.list[[i]] <- FindVariableFeatures(object = smp.list[[i]], selection.method = "vst", 
                                           nfeatures = 2000, verbose = FALSE)
  } else {
    smp.list[[i]] <- SCTransform(smp.list[[i]],vars.to.regress = corrections,verbose = T)
  }
}
```

## Integration CT and RA
```{r}
smp.anchors <- FindIntegrationAnchors(object.list = smp.list, dims = 1:opt$num_dim_CCA)

if(opt$reusePca) {
  print("Re use pca")
  smp.combined <- IntegrateData(anchorset = smp.anchors, weight.reduction = "pca", dims = 1:opt$num_dim_weight)
}

smp.combined <- IntegrateData(anchorset = smp.anchors, dims = 1:opt$num_dim_weight)
```


# Run the standard workflow for visualization and clustering
#Here certainly need to reuse SCTransform on combined data if norm_method = sct

```{r}
classicSeuratWorkflow <- function(smp.combined, corrections,outdir) {
  dir.create(outdir,recursive = T)
  smp.combined <- ScaleData(object = smp.combined, verbose = T,vars.to.regress = corrections)
  smp.combined <- RunPCA(object = smp.combined, npcs = opt$num_dim_integrated, verbose = FALSE)
  
  
  png(paste(outdir,"/ElbowPlot.png",sep =""))
  ElbowPlot(object = smp.combined,ndims = opt$num_dim_integrated)
  dev.off()
  
  
  # t-SNE UMAP and Clustering
  smp.combined <- RunUMAP(object = smp.combined, reduction = "pca", dims = 1:opt$num_dim)
  smp.combined <- RunTSNE(object = smp.combined, reduction = "pca", dims = 1:opt$num_dim)
  smp.combined <- FindNeighbors(object = smp.combined, reduction = "pca", dims = 1:opt$num_dim)
  smp.combined <- FindClusters(smp.combined, resolution = c(0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1,1.2))
  
  colPrefix <- paste0(DefaultAssay(smp.combined),"_snn_res.")
  
  Idents(smp.combined) <- paste(colPrefix,opt$resolution,sep = "")
  
  smp.combined@meta.data$AGE <- "Old"
  smp.combined@meta.data$AGE[which(smp.combined@meta.data$age == "2_months")] <- "Young"
  
  
  
  # Visualization of the samples
  p1 <- DimPlot(object = smp.combined, reduction = "umap", group.by = "sampleName")
  p2 <- DimPlot(object = smp.combined, reduction = "umap", label = TRUE)
  p3 <- DimPlot(object = smp.combined, reduction = "umap", group.by = "AGE")
  p4 <- DimPlot(object = smp.combined, reduction = "umap", group.by = "runDate")
  p3tsne <- DimPlot(object = smp.combined, reduction = "tsne", group.by = "AGE")
  p4tsne <- DimPlot(object = smp.combined, reduction = "tsne", group.by = "runDate")
  
  
  
  png(paste(outdir,"/umap_samples.png",sep =""),height = 800,width=1200)
  grid.arrange(p1, p2,ncol = 2)
  dev.off()
  
  png(paste(outdir,"/AGE.png",sep =""),height = 800,width=1200)
  grid.arrange(p3, p3tsne,ncol = 2)
  dev.off()
  
  png(paste(outdir,"/runDate.png",sep =""),height = 800,width=1200)
  grid.arrange(p4, p4tsne,ncol = 2)
  dev.off()
  
  png(paste(outdir,"/UmapClusters.png",sep =""),height = 800,width=1200)
  DimPlot(object = smp.combined, reduction = "umap", label = TRUE)
  dev.off()
  
  
  p1tsne <- DimPlot(object = smp.combined, reduction = "tsne", group.by = "sampleName")
  p2tsne <- DimPlot(object = smp.combined, reduction = "tsne", label = TRUE)
  
  png(paste(outdir,"/tsne_samples.png",sep =""),height = 800,width=1000)
  grid.arrange(p1tsne, p2tsne,ncol = 2)
  dev.off()
  
  #Visualistaoin one sample at a time
  sampleNames <- unique(smp.combined@meta.data$sampleName) 
  dir.create(paste(outdir,"/samples/umap/",sep = ""),recursive = T)
  dir.create(paste(outdir,"/samples/tsne/",sep = ""),recursive = T)
  
  #samplePlotList <- list()
  for (s in sampleNames) {
    png(paste(outdir,"/samples/umap/",s,".png",sep = ""))
    plot(DimPlot(object = smp.combined, cells= grep(s, smp.combined@meta.data$sampleName), reduction = "umap") + ggtitle(s))
    dev.off()
    png(paste(outdir,"/samples/tsne/",s,".png",sep = ""))
    plot(DimPlot(object = smp.combined, cells= grep(s, smp.combined@meta.data$sampleName), reduction = "tsne") + ggtitle(s))
    dev.off()
  }
  
  
  umapListRes <- list()
  tsneListRes <- list()
  for (r in c(0.6,0.8,1,1.2)) {
    umapListRes[[as.character(r)]] <- DimPlot(smp.combined,
                                              reduction = "umap",
                                              label = T,
                                              group.by = paste(colPrefix,r,sep="")) + 
      NoLegend() +
      ggtitle(paste("res",r))
    
    tsneListRes[[as.character(r)]] <- DimPlot(smp.combined,
                                              reduction = "tsne",
                                              label = T,
                                              group.by = paste(colPrefix,r,sep="")) + 
      NoLegend() +
      ggtitle(paste("res",r))
  }
  
  png(paste(outdir,"/tsne_different_res.png",sep = ""),height = 800,width = 800)
  grid.arrange(tsneListRes[[1]],tsneListRes[[2]],tsneListRes[[3]],tsneListRes[[4]])
  dev.off()
  
  png(paste(outdir,"/umap_different_res.png",sep = ""),height = 800,width = 800)
  grid.arrange(umapListRes[[1]],umapListRes[[2]],umapListRes[[3]],umapListRes[[4]])
  dev.off()
  
  
  
  smp.combined@meta.data$numclust <- smp.combined@meta.data[,paste(colPrefix,opt$resolution,sep = "")]
  
  
  #Check for unwanted source of variation UMAP
  pPhases <- DimPlot(smp.combined,group.by = "phases")
  if (!is.null(smp.combined@meta.data$predicted)) {
    pPred <- DimPlot(smp.combined,group.by = "predicted")
  } else {
    pPred <-  DimPlot(smp.combined)
  }
  pUMI <- FeaturePlot(smp.combined, "nCount_RNA")
  pMito <- FeaturePlot(smp.combined, "percentMito")
  
  png(paste(outdir,"/umap_factors.png",sep = ""),height = 800,width = 800)
  grid.arrange(pPhases,pPred,pUMI,pMito)
  dev.off()
  
  #Check for unwanted source of variation tSNE
  pPhases <- DimPlot(smp.combined,group.by = "phases",reduction = "tsne")
  if (!is.null(smp.combined@meta.data$predicted)) {
    pPred <- DimPlot(smp.combined,group.by = "predicted",reduction = "tsne")
  } else {
    pPred <-  DimPlot(smp.combined,reduction = "tsne")
  }
  pUMI <- FeaturePlot(smp.combined, "nCount_RNA",reduction = "tsne")
  pMito <- FeaturePlot(smp.combined, "percentMito",reduction = "tsne")
  
  png(paste(outdir,"/tsne_factors.png",sep = ""),height = 800,width = 800)
  grid.arrange(pPhases,pPred,pUMI,pMito)
  dev.off()
  return(smp.combined)
}

```

## Integration and correction
```{r}
print("Seurat workflow without integration")
DefaultAssay(object = smp.combined) <- "RNA"
smp.combined <- NormalizeData(object = smp.combined)
smp.combined <- FindVariableFeatures(object = smp.combined,selection.method = "vst", nfeatures = 2000, verbose = T)

if (!is.null(corrections)) {
  print("Seurat workflow without correction")
  classicSeuratWorkflow(smp.combined,correction = NULL,outdir = paste0(outdir,"/SeuratMergingWoIntegrationWoCr/"))
}

classicSeuratWorkflow(smp.combined,corrections = corrections,outdir = paste0(outdir,"/SeuratMergingWoIntegration/"))



print("Seurat workflow with integration")

DefaultAssay(object = smp.combined) <- "integrated"


if (!is.null(corrections)) {
  print("Seurat workflow without correction")
  classicSeuratWorkflow(smp.combined,correction = NULL,outdir = paste0(outdir,"/SeuratWithoutCr/"))
}

smp.combined <- classicSeuratWorkflow(smp.combined,corrections = corrections,outdir =outdir)
```

## Age Sample and runDte for each cluster
```{r}
#AGE prop for each cluster 
age_clust <- getAGEPropPerClustBarplot(smp.combined)

#sample prop for each cluster
sample_clust <- getSamplePropPerClustBarplot(smp.combined)

#runDate prop for each cluster
runDate_clust <- getRunDatePropPerClustBarplot(smp.combined)

png(paste(outdir,"/propPerClust.png",sep = ""))
grid.arrange(age_clust,sample_clust,runDate_clust,
             DimPlot(object = smp.combined, reduction = "tsne", label = TRUE))
dev.off()

```

## saving results

#saveRDS(smp.combined,paste(outdir,"/combined.rds",sep =""))
